AppBar can now have custom titleHeight

RaviKavaiya opened the pull request:
#### NOTE: This PR is same as #48557, but fixes identified changes. That was closed. Still, any suggestions are welcome.

Allows `AppBar` (and `SliverAppBar`) to have any height we specify.

## Description


- Previously, `AppBar`'s height was limited to 56.0 using a constant `kToolbarHeight`. This allows us to change AppBar's height using `titleHeight` parameter. If we omit this parameter, it fallbacks to `kToolbarHeight` (the default behavior), otherwise acts accordingly.

- When we specify `titleHeight`, the icons (`leading` and `actions`) are also centered vertically. (the default behavior). So, I also introduced a boolean `centerIcons` parameter, which when set to `false`, the icons stay on top (see below screenshots for more clarity).

### <a href="//imgur.com/a/TqmuR4x">SCREENSHOTS</a>

## Related Issues

#### using `titleHeight` parameter:
#23373 solved
#40134 solved
#7330 looks related

#### using `centerIcons` parameter:
- No issue is related with this parameter

## Tests

I added the following tests:

- Added test in `app_bar_test.dart` so that we can check custom title height provided.
- Added test in `floating_action_button_location_test.dart` so that we can check mini-start top FAB location.
- Added test in `scrollable_semantics_test.dart` so that showOnScreen works with pinned app bar.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change.
   
<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

RaviKavaiya committed:
AppBar can now have custom titleHeight

Now, AppBar's height is not limited by 56.0 pixels. Using 'titleHeight', we can change this to any height we want.

dnfield commented on packages/flutter/lib/src/material/app_bar.dart:
```diff
@@ -1087,6 +1123,20 @@ class SliverAppBar extends StatefulWidget {
+  /// If false, icons will be placed on top.
+  ///
+  /// Defaults to true.
+  final bool centerIcons;
```
I think we should hold off on this and wait for someone to ask for it.

I'm having trouble imagining why you would want a title bar where the actions were not centered along with the text. We should generally avoid adding API until we're sure we need it.

dnfield replied:
(And in fact, I'm not sure a simple bool even makes the most sense here - we'd probably want some kind of alignment based property instead if we did this)

dnfield commented on packages/flutter/lib/src/widgets/navigation_toolbar.dart:
```diff
@@ -58,6 +61,24 @@ class NavigationToolbar extends StatelessWidget {
+  /// If false, icons will be placed on top.
+  ///
+  /// Defaults to true.
+  final bool centerIcons;
```
Ditto.

dnfield commented on packages/flutter/lib/src/widgets/navigation_toolbar.dart:
```diff
@@ -32,8 +32,11 @@ class NavigationToolbar extends StatelessWidget {
     this.trailing,
     this.centerMiddle = true,
     this.middleSpacing = kMiddleSpacing,
+    this.defaultToolbarHeight,
```
Why are we calling this `defaultToolbarHeight` instead of just `toolbarHeight`?

dnfield replied:
I think we should probably ditch this along with centerIcons for now, or move it to a separate PR.

dnfield reviewed (commented):
I think the centerIcons/defaultToolbarHeight properties should probably be dropped from this PR.  It's not clear what the use cases for them would be yet.  It also doesn't have any tests

RaviKavaiya committed:
Removed unnecessary parameters

The parameters 'centerIcons' and 'defaultToolbarHeight' were removed as they were not necessary.

RaviKavaiya commented:
@dnfield  Removed unnecessary parameters

dnfield reviewed (commented):
This looks good to me - someone more familiar with Material Design is going to take a second pass at it soon!

Thanks for sticking with it!

AntiqPeace commented:
@RaviKavaiya  @dnfield any update on this?

HansMuller commented on packages/flutter/lib/src/material/app_bar.dart:
```diff
@@ -27,16 +27,18 @@ const double _kLeadingWidth = kToolbarHeight; // So the leading button is square
 
 // Bottom justify the kToolbarHeight child which may overflow the top.
```
This comment will need to be updated

RaviKavaiya replied:
Sorry, I did not understand.

HansMuller replied:
The child isn't "kToolbarHeight" any more

HansMuller commented on packages/flutter/lib/src/material/app_bar.dart:
```diff
@@ -27,16 +27,18 @@ const double _kLeadingWidth = kToolbarHeight; // So the leading button is square
-  const _ToolbarContainerLayout();
+  const _ToolbarContainerLayout(this.titleHeight) : assert(titleHeight != null);
+
+  final double titleHeight;
```
This isn't just the height of the title, it's the height of the toolbar (the NavigationToolBar built by the AppBar). It would be better to call it toolbarHeight. That would also help one understand why the default is kToolbarHeight :-). The docs for toolbarHeight should make its effect on leading/title/actions clear.

RaviKavaiya replied:
Okay

HansMuller commented on packages/flutter/lib/src/material/app_bar.dart:
```diff
@@ -194,15 +196,17 @@ class AppBar extends StatefulWidget implements PreferredSizeWidget {
     this.primary = true,
     this.centerTitle,
     this.titleSpacing = NavigationToolbar.kMiddleSpacing,
+    this.titleHeight = kToolbarHeight,
```
Would prefer that the default value for this parameter was null. That leaves open the possibility of adding toolbarHeight to AppBarTheme.

RaviKavaiya replied:
In #48557, I did the same thing. But as suggested by @dnfield , I did it this way. This also looks clean and easy to understand.

dnfield replied:
@HansMuller is more familiar with the Material system and theming - I would go with what he says here.

dnfield replied:
Sorry for causing any confusion or extra work!

HansMuller commented:
@RaviKavaiya it's been over a month and it looks like you haven't had time to update the PR per review feedback (like https://github.com/flutter/flutter/pull/48621#discussion_r373756461). I'm going to close it now. If you decide to resume working on it, feel free to reopen the PR.
